java/com/google/security/wycheproof/testcases/AesGcmTest.java

```diff
@@ -268,6 +270,169 @@ public void testByteBufferAlias() throws Exception {
+   * make the backing array visible.
+   *
+   * Note that bugs in this often require a sizeable input to reproduce; the default implementation of 
+   * engineUpdate(ByteBuffer, ByteBuffer) copies through 4KB bounce buffers, so we need to use something larger to see
+   * any problems - 8KB is what we use here.
+   *
+   * @see https://bugs.openjdk.java.net/browse/JDK-8181386
+   */
+  public void testByteBufferShiftedAlias() throws Exception {
+    byte[] ptVector = new byte[8192];
+
+    for (int i = 0; i < 3; i++) {
+      // outputOffset = offset relative to start of input.
+      for (int outputOffset = -1; outputOffset <= 1; outputOffset++) {
+        SecretKeySpec key = new SecretKeySpec(new byte[16], "AES");
+        GCMParameterSpec parameters = new GCMParameterSpec(128, new byte[16]);
+        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
+        cipher.init(Cipher.ENCRYPT_MODE, key, parameters);
+
+        ByteBuffer output, input, inputRO;
+
+        // We'll try three scenarios: Ordinary array backed buffers, array backed buffers where one is read-only, and
+        // direct byte buffers.
+        String mode;
+        // offsets relative to start of buffer
+        int inputOffsetInBuffer = 1;
+        int outputOffsetInBuffer = inputOffsetInBuffer + outputOffset;
+        int sliceLength = cipher.getOutputSize(ptVector.length);
+        int bufferSize = sliceLength + Math.max(inputOffsetInBuffer, outputOffsetInBuffer);
+        switch (i) {
+          case 0:
+          case 1: {
+            byte[] buffer = new byte[bufferSize];
+            // It's important to slice() here as otherwise later when we flip() position will be reset to 0.
+            output = ByteBuffer.wrap(buffer, outputOffsetInBuffer, sliceLength).slice();
+            input = ByteBuffer.wrap(buffer, inputOffsetInBuffer, sliceLength).slice();
+
+            if (i == 1) {
+              mode = "array backed buffers with RO buffer";
+              inputRO = input.asReadOnlyBuffer();
+            } else {
+              mode = "array backed buffers";
+              inputRO = input.duplicate();
+            }
+
+            break;
+          }
+          case 2: {
+            mode = "direct buffers";
+            ByteBuffer buf = ByteBuffer.allocateDirect(bufferSize);
+            output = buf.duplicate();
+            output.position(outputOffsetInBuffer);
+            output.limit(sliceLength + outputOffsetInBuffer);
+            output = output.slice();
+
+            input = buf.duplicate();
+            input.position(inputOffsetInBuffer);
+            input.limit(sliceLength + inputOffsetInBuffer);
+            input = input.slice();
+
+            inputRO = input.duplicate();
+            break;
+          }
+          default: {
+            throw new AssertionError("Unknown test index " + i);
+          }
+        }
+
+        // Now early points to all but the last byte of the buffer, and late and lateRead both point to all but the
```

Comment:
Oops, I think that's left over from before I refactored this code. I'll rewrite these comments.